Add new type '#' (number) for cmdline and use it in get_number

coditva opened the pull request:
See #7857

coditva committed:
Make inputlist execute with cmdline_prompt

See #7857

UtkarshMe commented on src/nvim/misc1.c:
```diff
@@ -2408,10 +2408,37 @@ get_number (
+  cmd_silent = false;  // Want to see the prompt.
+  cmdline_row = msg_row;
+
+  select = (char_u *)getcmdline_prompt('@', "", 0, xp_type, xp_arg,
```
This puts the cursor in new line. How can we avoid that?

bfredl replied:
@UtkarshMe The text that is to be displayed on the same line should be passed here as the second argument. You will need to add an argument to `get_number` so `prompt_for_number` and `ml_recover` can send this string in.

justinmk commented:
cc @ZyX-I @bfredl

coditva committed:
Add prompt in get_number() and add '#' mode

- Modify get_number() to also take a string as a prompt message.
- Add a '#' (number for input()) mode in command_line_enter which does not
does not echo alphabets when taking input in command line.

UtkarshMe commented:
One test is failing: "`NULL list is accepted as an empty list by inputlist()`"
I'm unable to figure out why...

bfredl commented on src/nvim/ex_getln.c:
```diff
@@ -5289,6 +5300,7 @@ static HistoryType hist_char2type(const int c)
     case '=': {
       return HIST_EXPR;
     }
+    case '#':
```
should either be its own history or `HIST_INVALID`

bfredl commented:
It's because that line is not echoed as a message anymore. It is unrelated to what the test tests (that nvim doesn't crash or behave weirdly with NULL list), so just update the test to the new behavior.

bfredl commented on src/nvim/ex_getln.c:
```diff
@@ -1005,6 +1009,12 @@ static int command_line_execute(VimState *state, int key)
   s->do_abbr = true;             // default: check for abbreviation
+
+  // if in number mode, key is alphabet
+  if (s->firstc == '#' && ASCII_ISALPHA(s->c)) {
```
isn't it better to check if it is a digit?

bfredl replied:
well, then this check should be moved to after special chars are checked, and just before a char is inserted into the cmdline string.

bfredl commented on src/nvim/ex_getln.c:
```diff
@@ -320,7 +323,8 @@ static uint8_t *command_line_enter(int firstc, long count, int indent)
 
-  if (s->firstc == '/' || s->firstc == '?' || s->firstc == '@') {
+  if (s->firstc == '/' || s->firstc == '?' || s->firstc == '@'
+      || s->firstc == '#') {
```
I don't think langmap should be invoked (so remove this).

bfredl commented on src/nvim/ex_getln.c:
```diff
@@ -263,7 +263,10 @@ static uint8_t *command_line_enter(int firstc, long count, int indent)
   s->old_botline = curwin->w_botline;
 
   // set some variables for redrawcmd()
-  ccline.cmdfirstc = (s->firstc == '@' ? 0 : s->firstc);
```
The existing shenanigans with `cmdfirstc` is a bit unfortunate... `get_cmdline_type()` should support the new `#` type. Either a boolean flag for this type needs to be added in `ccline`, or perhaps better variable that is the unadulterated `s->firstc` value (to support future expansion), which probably also should be used in `get_histtype`

UtkarshMe replied:
Ok, why isn't `ccline.cmdfirstc = s->firstc` simply? Why make it 0 for '@'?

UtkarshMe replied:
Got it. This has to be 0 for both '@' and '#' because it is the first character shown on the command line.

So we keep it 0 for both
**OR** we can change all the `cmdfirstc == NUL` checks.

coditva committed:
Add hist_type for '#' mode and postpone digit check

coditva committed:
Correct test for inputlist

coditva committed:
Fix lint errors

coditva committed:
Add a new history table 'number'

UtkarshMe commented:
@bfredl I need help... the log messages say that legacy tests are failing but `make oldtest` is successful on my system (`ALL DONE`).

Also, QuickBuild has these two failures for `functionaltest-macosx-64-debug`:
- timers can be stopped
- timers can be stopped from the handler
functional tests pass on my system (`Linux 4.15.5-1-ARCH`).

bfredl commented:
Timer tests are known to be fragile, can ignore them for now. Though with oldtests, at least the ASAN built it looks like it has detected a memory leak.

coditva committed:
Fix memory leak

coditva committed:
Fix lint errors

coditva committed:
Don't invoke langmap for '#' type

bfredl commented on test/functional/ui/cmdline_spec.lua:
```diff
@@ -168,7 +168,7 @@ describe('external cmdline', function()
       eq({{
         content = { { {}, "default" } },
-        firstc = "",
+        firstc = "@",
```
I'm not sure this change is good or not. IIRC emtpy firstc should only occur with `input()` so there is no strong reason to change it (and if not, we should consider setting firstc in that other place instead)

Also, in this file a test should be added that `inputlist()` works as expected with external cmdline (firstc is set to `#`, prompt if applicable)

bfredl replied:
Then also, we should consider allowing GUI:s to convert `inputlist()` in to proper dialogs. That would be a separate follow up PR though (so no action needed here)

UtkarshMe replied:
I think we already use `@` for setting the mode for `input()` function in `getcmdline()` and the likes, so while adding `#` which also uses `input()` function, it seemed better to use `@` as `firstc` to distinguish between the two.

bfredl replied:
Not sure i follow, only `input()` is `input()` (well, except for `inputdialog()` for the moment). Number (`inputlist()` ) is a new and different cmdline type, and should definitely be marked with `firstc = '#'` (even if prompt is available, we will need to update the docs). I was speaking if some other cmdline type  also happened to set `firstc = ""` already (maybe some ex mode state), it is probably much rarer and so nicer to clients to change that instead.

UtkarshMe replied:
Oh, I get it now! There is another mode `:insert` with `firstc = NUL`. The line `ccline.cmdfirstc = (firstc == '@' ? 0 : firstc)` makes `cmdfirstc` for both `:insert` and `input()` same, but I guess that's because they are so similar.
I think I understand it now and should revert back to `firstc = ""` for `@` mode.

bfredl replied:
Yes `:append` /`:insert` should maybe be marked with `-`, but overall ex mode commands will not look well with ext-ui so this could be considered in a later PR (where "later" could mean "never", if the intersection of users using ex mode and extended UI is empty, which is not impossible)

coditva committed:
Revert changing firstc value for input()

UtkarshMe commented:
TODO:
- [ ] Add UI ext-cmdline test for `inputlist()`

coditva committed:
Add UI test for inputlist()

bfredl commented on test/functional/ui/cmdline_spec.lua:
```diff
@@ -187,6 +187,36 @@ describe('external cmdline', function()
+      {1:~                        }|
+      input0                   |
+      input1                   |
+            ^                   |
```
Hmm cursor position is weird, ideally it should be on the first column, or perhaps avoid the last blank line altogether if possible.

UtkarshMe replied:
I think this code is responsible for putting the last line as well as the cursor position:
```c
// from line 11279 in eval.c
TV_LIST_ITER_CONST(argvars[0].vval.v_list, li, {
    msg_puts(tv_get_string(TV_LIST_ITEM_TV(li)));
    msg_putchar('\n');
  });
```
Removing the newline would be complicated... Also, how do we change the cursor position? It stays at the position that `msg_putchar()` or `msg_puts()` leaves it.

bfredl replied:
I think you can use `TV_LIST_ITEM_NEXT(list, li) != NULL` to check if an item is last and then avoid the `msg_putchar('\n')`. ( when `ui_is_external(kUICmdline)` is true)

Alternatively, i think you can just add an `msg_putchar('\r')` at the end to move it back to first column.

ZyX-I replied:
Better to check against equality to `tv_list_last()`: it makes the intention more clear.

UtkarshMe replied:
After rebasing with master, if `msg_putchar('\n')` is not present after every list item, the list isn't displayed correctly. Should I revert?

bfredl commented on src/nvim/misc1.c:
```diff
@@ -2408,10 +2412,24 @@ get_number (
-  for (;; ) {
-    ui_cursor_goto(msg_row, msg_col);
-    c = safe_vgetc();
+  while (select) {
```
Most handling of special chars can be removed from this loop.

We probably still want to support mouse in this mode. Look at `case K_MOUSELEFT` in `command_line_handle_key`, where `#` mode should be special cased. (replacing the text with the number printed as a string is fine)

UtkarshMe replied:
I removed the redundant handling, but how do I make the `#` mode special case in `command_line_handle_key`? Do I add a condition in this:
```c
if (!mouse_has(MOUSE_COMMAND)) {
  return command_line_not_changed(s);
}
```

bfredl replied:
Good question, previously mouse works in `inputlist()` if the mode from which it was called was included in `set mouse`, but now it will be controlled by if `mouse=c` (or `a`, which is the default) is included. We could add an exception so that mouse is always active with `firstc='#'` unless `mouse` is completely empty. (this will also require change to `setmouse()`). For now, you can start with it working with `mouse=c` or `a`, so put the condition _after_ this if block.

if `s->firstc == '#'` you will need to check if the mouse is above the cmdline row, and then `return 0` to abort the cmdline. But you need to somehow signal back to  `get_number` that mouse was used (so that it reads `mouse_row` instead), as `getcmdline_prompt` is not called much, adding a `bool *` out arg would be fine.

UtkarshMe replied:
Hi, sorry for the late reply (acads in college), how do I test if mouse is working? I tried to use mouse to select options in Kitty, Gnome Terminal and Tilda but it did not work in any (even in the master branch)... Which terminal can I use to test mouse?

bfredl replied:
What does `set mouse` give? Try `set mouse=a` . The two first I know should work, if still not, what is `$TERM` ?

UtkarshMe replied:
Ok! So so sorry for this... I had not set mouse. :sweat_smile:

coditva committed:
Remove the last empty line in inputlist items

coditva committed:
Remove redundant handling of chars from getnumber

coditva committed:
Allow using mouse in `#` mode

UtkarshMe commented:
The test `test/functional/ui/cmdline_spec.lua` passes locally but not on Travis.
This is the error:
```
Expected:
  |*{1:~                        }|
  |{1:~                        }|
  |{1:~                        }|
  |input0                   |
  |input1^                   |
Actual:
  |*                         |
  |{1:~                        }|
  |{3:                         }|
  |input0                   |
  |input1^                   |
```

bfredl commented:
Please rebase on master and try again locally, there was a change on master how multiline messages are displayed, this looks like it could be it.

bfredl commented on src/nvim/ex_getln.c:
```diff
@@ -214,6 +215,9 @@ static int hislen = 0;                  /* actual length of history tables */
 static bool getln_interrupted_highlight = false;
 
+/// Pointer to flag which keeps track if mouse was used
+static int *mouse_used_p = NULL;
```
I think we should avoid storing pointers if not necessary. Instead set the return value in `getcmdline_prompt` after `getcmdline`. Also place the value inside `ccline`, not globally. (if mouse support is conditional, an explicit bool variable is also better than a NULL check far from where the pointer is defined)

coditva committed:
Put global pointer for mouse_used in ccline

coditva committed:
Revert 2c25b9f: Remove the last line in inputlist

coditva committed:
Fix lint issues

bfredl commented on src/nvim/ex_getln.c:
```diff
@@ -2013,13 +2029,17 @@ char *getcmdline_prompt(const char firstc, const char *const prompt,
   ccline.xp_context = xp_context;
   ccline.xp_arg = (char_u *)xp_arg;
-  ccline.input_fn = (firstc == '@');
+  ccline.input_fn = (firstc == '@' || firstc == '#');
```
I don't think this should be added (it seems to mostly used for expansion)

bfredl commented on src/nvim/misc1.c:
```diff
@@ -2408,34 +2412,37 @@ get_number (
+  cmdline_row = msg_row;
+
+  select = getcmdline_prompt('#', prompt, 0, EXPAND_NOTHING, NULL,
+                             input_callback, mouse_used);
```
can use CALLBACK_NONE in arg list and drop `input_callback`

bfredl commented on src/nvim/ex_getln.h:
```diff
@@ -36,6 +36,7 @@ typedef enum {
   HIST_SEARCH,        ///< Search commands.
   HIST_EXPR,          ///< Expressions (e.g. from entering = register).
   HIST_INPUT,         ///< input() lines.
+  HIST_INPUT_NUM,     ///< input() numbers.
```
inputlist

bfredl commented on src/nvim/ex_getln.c:
```diff
@@ -1477,6 +1480,11 @@ static int command_line_handle_key(CommandLineState *s)
       return command_line_not_changed(s);                   // Ignore mouse
     }
 
+    if (s->firstc == '#' && mouse_row < cmdline_row) {
```
We probably only want to do this for press events, not DRAG or RELEASE (it should then fallthrough to the ordinary code below)

UtkarshMe replied:
Something like this?
```c
if (s->firstc == '#' && mouse_row < cmdline_row && s->ignore_drag_release)
```
It does solve the problem of returning as soon as mouse is pressed. It returns after mouse is released.

bfredl replied:
I don't think that is a problem, activation on press is expected. The situation is rather to press mouse at cmdline (which moves the cursor), and then accidentially drag over the second last line, which should continue to move the cursor, not accept the last choice.

UtkarshMe replied:
After the change mentioned above, if you drag over the second last line from the cmdline, it won't be accepted but if you release while on it, it will be. Is this the expected behavior?

bfredl replied:
hmm I will try it

bfredl replied:
To me it feels a bit weird, better to introduce `bool is_press = s->c == K_LEFTMOUSE || s->c == K_RIGHTMOUSE` and use it.

bfredl commented:
Seems to work pretty well now. A bit more tests would be good (without ext_cmdline, can be in `ui/input_spec.lua`), for instance:
- type number and CR
- type number and ESC
- type alphabetical chars
- mouse usage (in tests use `feed('<LeftMouse><col,row>')`, see `ui/mouse_spec.lua` for examples)

coditva committed:
Fix comments and remove redundant variables

coditva committed:
Exit cmdline only if mouse is pressed

bfredl commented:
Actually, `eval/inputlist_spec.lua` would make most sense.

coditva committed:
Add test for inputlist

UtkarshMe commented:
`:h inputlist` shows that "clicking on the prompt one more than the length is returned".
But in this pull request, clicking on prompt does nothing.
Should I change `mouse_row < cmdline_row` to `mouse_row <= cmdline_row` (maybe remove it altogether?) or is this behavior better?

bfredl commented:
Probably best to preserve existing behavior of `inputlist()`, so remove the check.

coditva committed:
Return items length + 1 on clicking prompt

coditva committed:
Remove unused functions in tests

coditva committed:
Remove unused variable in test

justinmk commented:
This test is failing in multiple builds:

```
1 FAILED:
Found errors in Test_cscopeWithCscopeConnections():
function RunTheTest[35]..Test_cscopeWithCscopeConnections line 206: Expected '"cs add find help kill reset show' but got 'cs find s main'
```

You can run it locally like this:

    TEST_FILE=test_cscope.res make oldtest

justinmk commented:
- Can you explain how this affects https://github.com/neovim/neovim/issues/7857 ?
- Need to update docs (`:help inputlist()`, at least).

UtkarshMe commented:
- Before this, `f_inputlist()` used to execute without getting into `state_enter()` but directly calling `safe_vgetc()`. This, I think, made it a blocking function and timers were not processed.
- I introduced a new mode `#` (number) for `struct cmdline_info` which takes input from the commandline but only digits are processed and everything else (alphabets, special chars) is rejected. This is ideal for this scenario and can be extended to implement any new function (e.g. `getnumber()`) in the future which only reads numbers.
- There's no change in `inputlist()` behavior so, `:h inputlist` need not be updated.

justinmk commented:
Ok, so there's no user-facing enhancement here?

UtkarshMe commented:
No.

justinmk commented:
Is this still WIP, or potentially ready for merge?

UtkarshMe commented:
Commenting out this section solves the [problem](https://github.com/neovim/neovim/pull/8076#issuecomment-383284481):
```viml
" From line 192 in testdir/test_cscope.vim

" Test 19: this should trigger call to cs_print_tags()
" Unclear how to check result though, we just exercise the code.
set cst cscopequickfix=s0
call feedkeys(":cs find s main\<CR>", 't')


```
Or even changing the last line to:
```viml
call feedkeys(":cs find s int\<CR>", 't')
```
@justinmk, I am unsure how to this is related to my PR and how to proceed with it.

justinmk commented:
cscope was removed, that might unblock this.

zeertzjq commented:
This seems superseded by #31525, but maybe some parts of this PR are still useful?

justinmk commented:
Lots of code and reviews spent on this, but still wondering what problem it solves.
